src/rust/engine/src/watch.rs

```diff
@@ -0,0 +1,190 @@
+          // The Graph has been dropped: we're done.
+          break;
+        };
+        match event_res {
+          Ok(Ok(ev)) => {
+            let paths: HashSet<_> = ev
+              .paths
+              .into_iter()
+              .map(|path| {
+                // relativize paths to build root.
+                if path.starts_with(&build_root) {
+                  path.strip_prefix(&build_root).unwrap().into()
+                } else {
+                  path
+                }
+              })
+              .collect();
+            warn!("notify invalidating {:?} because of {:?}", paths, ev.kind);
+            InvalidationWatcher::invalidate(&graph, &paths, "notify");
+          }
+          Ok(Err(err)) => {
+            if let notify::ErrorKind::PathNotFound = err.kind {
+              warn!("Path(s) did not exist: {:?}", err.paths);
+              continue;
+            } else {
+              error!("File watcher failing with: {}", err);
+              break;
+            }
+          }
+          Err(RecvTimeoutError::Timeout) => continue,
+          Err(RecvTimeoutError::Disconnected) => {
+            // The Watcher is gone: we're done.
+            break;
+          }
+        };
+      }
+      warn!("Watch thread exiting.");
+      // Signal that we're exiting (which we would also do by just dropping the channel).
+      let _ = thread_liveness_sender.send(());
+    });
+
+    Ok(InvalidationWatcher {
+      watcher,
+      liveness: thread_liveness_receiver,
+    })
+  }
+
+  ///
+  /// Watch the given path non-recursively.
+  ///
+  pub fn watch(&self, path: PathBuf) -> BoxFuture<(), ()> {
+    WatchFuture::new(self.watcher.clone(), path.clone())
+      .map_err(move |e| warn!("Failed to watch path {:?}, with error {:?}", path, e))
+      .to_boxed()
+  }
+
+  ///
+  /// Returns true if this InvalidationWatcher is still valid: if it is not valid, it will have
+  /// already logged some sort of error, and will never restart on its own.
+  ///
+  pub fn running(&self) -> bool {
+    match self.liveness.try_recv() {
+      Ok(()) | Err(TryRecvError::Disconnected) => false,
+      Err(TryRecvError::Empty) => true,
+    }
+  }
+
+  pub fn invalidate(graph: &Graph<NodeKey>, paths: &HashSet<PathBuf>, caller: &str) -> usize {
+    let InvalidationResult { cleared, dirtied } = graph.invalidate_from_roots(move |node| {
+      if let Some(fs_subject) = node.fs_subject() {
+        paths.contains(fs_subject)
+      } else {
+        false
+      }
+    });
+    // TODO: The rust log level is not currently set correctly in a pantsd context. To ensure that
+    // we see this even at `info` level, we set it to warn. #6004 should address this by making
+    // rust logging re-configuration an explicit step in `src/python/pants/init/logging.py`.
+    warn!(
+      "{} invalidation: cleared {} and dirtied {} nodes for: {:?}",
+      caller, cleared, dirtied, paths
+    );
+    cleared + dirtied
+  }
+}
+
+struct WatchFuture {
```

Comment:
So, rather than a custom `Future` type, it would be preferable to use https://github.com/pantsbuild/pants/blob/992a4ca82863be15efa48059abb7ad2f0231c599/src/rust/engine/task_executor/src/lib.rs#L129-L149 , which has the advantage that it pools threads from:
1. the caller of `watch`.
2. the implementation of `watch` (using a clone of the `Executor` on `self`: it's cheaply cloneable).